File: apps/api/src/modules/integrations/application/game-sync.service.ts
import { Injectable } from '@nestjs/common';
import type { ImportGameResultDto } from '@gamescore/types';

import { NotFoundError } from '../../../common/errors/app.exception';
import { ERROR_CODES } from '../../../common/errors/error-codes';
import { PrismaService } from '../../../common/prisma/prisma.service';
import { GameRepository } from '../../games/repositories/game.repository';
import { IgdbClient } from '../igdb/igdb.client';
import { GameImportService } from './game-import.service';

/**
 * Re-applies IGDB data to an existing game, skipping any field a human has
 * edited. That is what `Game.editedFields` is for.
 */
@Injectable()
export class GameSyncService {
  constructor(
    private readonly igdb: IgdbClient,
    private readonly importer: GameImportService,
    private readonly games: GameRepository,
    private readonly prisma: PrismaService,
  ) {}

  async sync(gameId: string): Promise<ImportGameResultDto> {
    const game = await this.games.findById(gameId);
    if (!game) throw new NotFoundError(ERROR_CODES.GAME_NOT_FOUND, 'Game not found');

    const source = await this.prisma.gameExternalSource.findFirst({
      where: { gameId, provider: 'IGDB' },
    });
    if (!source) {
      throw new NotFoundError(ERROR_CODES.IGDB_GAME_NOT_FOUND, 'This game has no IGDB source');
    }

    const payload = await this.igdb.getGameById(source.externalId);
    if (!payload) {
      throw new NotFoundError(ERROR_CODES.IGDB_GAME_NOT_FOUND, 'IGDB game not found');
    }

    const warnings: string[] = [];
    const mapped = this.importer.map(payload, warnings);
    const edited = new Set(game.editedFields);

    const next = {
      name: edited.has('name') ? game.name : mapped.name,
      summary: edited.has('summary') ? game.summary : mapped.summary,
      description: edited.has('description') ? game.description : mapped.description,
      developer: edited.has('developer') ? game.developer : mapped.developer,
      publisher: edited.has('publisher') ? game.publisher : mapped.publisher,
      releaseDate: edited.has('releaseDate') ? game.releaseDate : mapped.releaseDate,
      coverImageUrl: edited.has('coverImageUrl') ? game.coverImageUrl : mapped.coverImageUrl,
      bannerImageUrl: edited.has('bannerImageUrl') ? game.bannerImageUrl : mapped.bannerImageUrl,
      trailerYoutubeId: edited.has('trailerYoutubeId') ? game.trailerYoutubeId : mapped.trailerYoutubeId,
      galleryImageUrls: edited.has('galleryImageUrls') ? game.galleryImageUrls : mapped.galleryImageUrls,
      editedFields: game.editedFields,
    };

    await this.games.updateEditorial(gameId, next);
    await this.prisma.gameExternalSource.update({
      where: { id: source.id },
      data: {
        lastSyncedAt: new Date(),
        externalUpdatedAt: payload.updated_at ? new Date(payload.updated_at * 1000) : source.externalUpdatedAt,
        rawPayload: payload as object,
      },
    });
    await this.importer.syncRelations(gameId, payload);

    return {
      gameId,
      slug: game.slug,
      name: next.name,
      provider: 'IGDB',
      externalId: source.externalId,
      created: false,
      warnings,
    };
  }
}

File: apps/api/src/GameScore.Infrastructure/Integrations/GameSyncService.cs
using GameScore.Application.Errors;
using GameScore.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;
using System.Text.Json;

namespace GameScore.Infrastructure.Integrations;

public sealed class GameSyncService
{
    private readonly IgdbClient _igdb;
    private readonly GameImportService _importer;
    private readonly GameScoreDbContext _db;

    public GameSyncService(IgdbClient igdb, GameImportService importer, GameScoreDbContext db)
    {
        _igdb = igdb;
        _importer = importer;
        _db = db;
    }

    public async Task<ImportResult> SyncAsync(Guid gameId, CancellationToken cancellationToken = default)
    {
        var game = await _db.Games.FirstOrDefaultAsync(g => g.Id == gameId, cancellationToken);
        if (game is null)
        {
            throw new NotFoundException(ErrorCodes.GameNotFound, "Game not found");
        }

        var source = await _db.GameExternalSources.FirstOrDefaultAsync(
            s => s.GameId == gameId && s.Provider == DbExternalProvider.IGDB,
            cancellationToken);
        if (source is null)
        {
            throw new NotFoundException(ErrorCodes.IgdbGameNotFound, "This game has no IGDB source");
        }

        var payload = await _igdb.GetGameByIdAsync(source.ExternalId, cancellationToken);
        if (payload is null)
        {
            throw new NotFoundException(ErrorCodes.IgdbGameNotFound, "IGDB game not found");
        }

        var warnings = new List<string>();
        var mapped = _importer.MapPayload(payload, warnings);
        var edited = game.EditedFields.ToHashSet(StringComparer.Ordinal);

        if (!edited.Contains("name"))
        {
            game.Name = mapped.Name;
        }

        if (!edited.Contains("summary"))
        {
            game.Summary = mapped.Summary;
        }

        if (!edited.Contains("description"))
        {
            game.Description = mapped.Description;
        }

        if (!edited.Contains("developer"))
        {
            game.Developer = mapped.Developer;
        }

        if (!edited.Contains("publisher"))
        {
            game.Publisher = mapped.Publisher;
        }

        if (!edited.Contains("releaseDate"))
        {
            game.ReleaseDate = mapped.ReleaseDate;
        }

        if (!edited.Contains("coverImageUrl"))
        {
            game.CoverImageUrl = mapped.CoverImageUrl;
        }

        if (!edited.Contains("bannerImageUrl"))
        {
            game.BannerImageUrl = mapped.BannerImageUrl;
        }

        if (!edited.Contains("galleryImageUrls"))
        {
            game.GalleryImageUrls = mapped.GalleryImageUrls;
        }

        game.UpdatedAt = DateTime.UtcNow;
        source.LastSyncedAt = DateTime.UtcNow;
        source.ExternalUpdatedAt = payload.Updated_at is null
            ? source.ExternalUpdatedAt
            : DateTimeOffset.FromUnixTimeSeconds(payload.Updated_at.Value).UtcDateTime;
        source.RawPayload = JsonSerializer.Serialize(payload);
        source.UpdatedAt = DateTime.UtcNow;

        await _db.SaveChangesAsync(cancellationToken);

        return new ImportResult(
            gameId,
            game.Slug,
            game.Name,
            Created: false,
            Provider: "IGDB",
            ExternalId: source.ExternalId,
            warnings);
    }
}
